src/workerd/api/streams/writable-sink-adapter.h

```diff
@@ -0,0 +1,446 @@
+  // or stored, but it is safe to return this from a getter multiple times
+  // to JavaScript as it will ensure that the same JS promise object is
+  // always returned until the backpressure state changes. This variation
+  // is not suitable for use within C++ code that needs to await on the
+  // ready promise because the internal jsg::Promise<void> object will
+  // no longer exist once the reference is passed out to JavaScript.
+  jsg::MemoizedIdentity<jsg::Promise<void>>& getReadyStable();
+
+  // Returns the options used to configure this adapter if the adapter
+  // is not closed or errored.
+  kj::Maybe<const Options&> getOptions();
+
+  void visitForGc(jsg::GcVisitor& visitor);
+  void visitForMemoryInfo(jsg::MemoryTracker& tracker) const;
+
+ private:
+  // Represents the active state of the adapter. Importantly, this state
+  // holds the both the underlying WritableStreamSink and the write queue.
+  // It must be held within an IoOwn.
+  struct Active;
+  struct Closed final {};
+  kj::OneOf<IoOwn<Active>, Closed, kj::Exception> state;
+
+  // Used for back pressure signaling. When backpressure is indicated, the
+  // readyResolver, ready, and readyWatcher will be replaced with a new set.
+  // When backpressure is relieved, the readyResolver will be resolved.
+  // The adapter will start out in a ready state.
+  class BackpressureState final {
+   public:
+    // Aborts backpressure signaling, likely because the adapter is being errored.
+    // Causes the ready promise to be rejected with the given reason.
+    void abort(jsg::Lock& js, const jsg::JsValue& reason);
+
+    // Releases backpressure, resolving the ready promise.
+    void release(jsg::Lock& js);
+
+    // Indicates that backpressure has been signaled and we are waiting
+    // for it to be released or aborted.
+    bool isWaiting() const;
+
+    // Returns a promise that resolves when backpressure is released.
+    // Note that every call to this returns a new jsg::Promise<void>
+    // instance. Callers that need a stable identity should use
+    // getReadyStable() instead (generally this is only the case when
+    // returning the promise to JavaScript via a getter).
+    jsg::Promise<void> getReady(jsg::Lock& js);
+
+    // Returns a memoized identity for the ready promise. This can be used
+    // to return a stable reference to the ready promise out to JavaScript
+    // that will not change identity between calls unless the backpressure
+    // state changes.
+    jsg::MemoizedIdentity<jsg::Promise<void>>& getReadyStable();
+
+   private:
+    kj::Maybe<jsg::Promise<void>::Resolver> readyResolver;
+    jsg::Promise<void> ready;
+    jsg::MemoizedIdentity<jsg::Promise<void>> readyWatcher;
+    BackpressureState(jsg::Promise<void>::Resolver&& resolver,
+        jsg::Promise<void>&& promise,
+        jsg::MemoizedIdentity<jsg::Promise<void>>&& watcher);
+
+    friend class WritableStreamSinkJsAdapter;
```

Comment:
Nit: just make this class a struct? It's defined as private to WritableStreamSinkJsAdapter, so I think it should be equivalent in terms of who has access.

Alternatively, if WritableStreamSinkJsAdapter only needs friend access for `newBackpressureState()` and `visitFor*()`, maybe move those three functions to BackpressureState itself.